cmake/std/trilinosprhelpers/setenvironment/test/test_ModuleHelper.py

```diff
@@ -0,0 +1,150 @@
+
+
+
+class mock_popen(object):
+    """
+    Abstract base class for popen mock
+    """
+    #__metaclass__ = abc.ABCMeta
+    def __init__(self, cmd, stdout=None, stderr=None):
+        print("mock_popen> {}".format(cmd))
+        self.stdout = stdout
+        self.stderr = stderr
+        self.returncode = None
+
+    #@abc.abstractmethod
+    def communicate(self):
+        print("mock_popen> communicate()")
+        stdout = b"stdout=1"
+        stderr = b"stderr=2"
+        self.returncode = 0
+        return (stdout,stderr)
+
+
+
+class mock_popen_status_ok(mock_popen):
+    """
+    Specialization of popen mock that will return with success.
+    """
+    def __init__(self, cmd, stdout=None, stderr=None):
+        super(mock_popen_status_ok, self).__init__(cmd,stdout,stderr)
+
+
+
+class mock_popen_status_error(mock_popen):
+    """
+    Specialization of popen mock that will return with success.
+    """
+    def __init__(self, cmd, stdout=None, stderr=None):
+        super(mock_popen_status_error, self).__init__(cmd,stdout,stderr)
+
+    def communicate(self):
+        print("mock_popen> communicate()")
+        stdout = b"stdout=1"
+        stderr = b"ERROR: Something wrong happened."
+        self.returncode = 1
+        return (stdout,stderr)
+
+
+class mock_popen_status_error_mlstatus(mock_popen):
+    """
+    Specialization of popen mock that will return with success.
+    """
+    def __init__(self, cmd, stdout=None, stderr=None):
+        super(mock_popen_status_error_mlstatus, self).__init__(cmd,stdout,stderr)
+
+    def communicate(self):
+        print("mock_popen> communicate()")
+        stdout = b"stdout=1"
+        stderr = b"_mlstatus = False"
+        self.returncode = 1
+        return (stdout,stderr)
+
+
+
+class moduleHelperTest(TestCase):
+    """
+    Main test driver for the module() function provided by the
+    ModuleHelper.py file
+    """
+    def setUp(self):
+        pass
+
+
+    @patch('subprocess.Popen', side_effect=mock_popen_status_ok)
+    def test_module_load_status_ok(self, arg_popen):
+        r = ModuleHelper.module("load", "sems-gcc/4.8.4")
+        print("result = {}".format(r))
+        self.assertEqual(0, r)
+
+
+    @patch('subprocess.Popen', side_effect=mock_popen_status_error)
+    def test_module_load_status_error(self, arg_popen):
+        r = ModuleHelper.module("load", "sems-gcc/4.8.4")
+        print("result = {}".format(r))
+        self.assertEqual(1, r)
+
+
+    @patch('subprocess.Popen', side_effect=mock_popen_status_ok)
+    def test_module_swap_status_ok(self, arg_popen):
+        r = ModuleHelper.module("swap", "sems-gcc/4.8.4", "sems-gcc/7.3.0")
+        print("result = {}".format(r))
+        self.assertEqual(0, r)
+
+
+    #@patch('subprocess.Popen', side_effect=mock_popen_status_ok)
+    def test_module_unload_status_ok(self):
+        with patch('subprocess.Popen', side_effect=mock_popen_status_ok):
+            r = ModuleHelper.module("unload", "sems-gcc/4.8.4")
```

Comment:
What's the difference between the `@patch` decorator and the `with patch(...):` pattern?

Reply:
They're different flavors of the same thing.